helix-view/src/document.rs

```diff
@@ -163,23 +172,107 @@ impl Document {
+        // context it is used.
+        let valid_up_to = encoding_rs::Encoding::utf8_valid_up_to(buf);
+        let max_buffer_len = decoder
+            .max_utf8_buffer_length(buf.len() - valid_up_to)
+            .unwrap();
+        valid_up_to + max_buffer_len
+    }
+
+    #[inline(always)]
+    fn calculate_encode_capacity(encoder: &mut encoding_rs::Encoder, buf: &[u8]) -> usize {
+        // Will never panic because the buffer size will never be close to usize::MAX in the
+        // context it is used.
+        let valid_up_to = encoding_rs::Encoding::utf8_valid_up_to(buf);
+        let max_buffer_len = encoder
+            .max_buffer_length_from_utf8_if_no_unmappables(buf.len() - valid_up_to)
+            .unwrap();
+        (valid_up_to + max_buffer_len).next_power_of_two()
+    }
+
     // TODO: async fn?
-    pub fn load(path: PathBuf) -> Result<Self, Error> {
-        use std::{fs::File, io::BufReader};
+    /// Open a file from a path, or create a new file if it doesn't exist.
+    pub fn open(path: PathBuf) -> Result<Self, Error> {
+        use std::{
+            fs::File,
+            io::{BufReader, Read},
+        };
 
-        let doc = if !path.exists() {
-            Rope::from("\n")
-        } else {
-            let file = File::open(&path).context(format!("unable to open {:?}", path))?;
-            let mut doc = Rope::from_reader(BufReader::new(file))?;
-            // add missing newline at the end of file
-            if doc.len_bytes() == 0 || doc.byte(doc.len_bytes() - 1) != b'\n' {
-                doc.insert_char(doc.len_chars(), '\n');
-            }
-            doc
+        if !path.exists() {
+            return Ok(Self::new());
+        }
+
+        // Create a 64kb buffer for reading the file into.
+        let mut buf = [0u8; 1 << 16];
+        let mut builder = RopeBuilder::new();
+        let mut file = BufReader::new(File::open(&path)?);
+        let (encoding, mut decoder, is_last) = {
+            let read = file.read(&mut buf)?;
+            let is_last = read < buf.len();
+            let mut encoding_detector = chardetng::EncodingDetector::new();
+            encoding_detector.feed(&buf, is_last);
+
+            let encoding = encoding_detector.guess(None, true);
+            let decoder = encoding.new_decoder();
+            builder.append(&encoding.decode(&buf[..read]).0);
+
+            (encoding, decoder, is_last)
         };
 
-        let mut doc = Self::new(doc);
+        // Check if the file was read to completion when we created the encoding.
+        if !is_last {
+            let capacity = Self::calculate_decode_capacity(&mut decoder, &buf);
+            // Creating a reusable buffer for writing the decoded text into.
+            // Theoretically it should lead to less allocations.
+            let mut buf_str = String::with_capacity(capacity);
+
+            // Loop through all the text in the file and decode it.
+            loop {
+                let read = file.read(&mut buf)?;
```

Comment:
I think before you do this read, you need to shift any bytes that weren't decoded in the previous iteration to the start of the buffer.  Then you'll want to append the new bytes to the end of that when doing this read.

The reason you need to do this is because the decoder actually isn't guaranteed to decode all the bytes in the buffer.  This can happen when, for example, the last byte in the buffer is part of a multi-byte character, the remaining bytes of which haven't been read from disk yet.

The `decode_to_string()` function (according to the documentation) returns:

> [...] a tuple that contains the `DecoderResult`, the number of bytes read and a boolean indicating whether replacements were done.

The "number of bytes read" is what you need to pay attention to for this.  Any bytes in the buffer beyond that read-count need to be shifted back to the start of the buffer for the next iteration.

Reply:
Also, I think we need some test cases that exercise this multi-chunk decoding process for a variety of text encodings.  And to that end, maybe the buffer doesn't really need to be so large.  Probably 4096 bytes is plenty, and then we can just have a few test files that are larger than that, in various different encodings.